libvast/src/system/connector.cpp

```diff
@@ -0,0 +1,256 @@
+    case caf::sec::remote_linking_failed:
+    case caf::sec::cannot_add_upstream:
+    case caf::sec::upstream_already_exists:
+    case caf::sec::invalid_upstream:
+    case caf::sec::cannot_add_downstream:
+    case caf::sec::downstream_already_exists:
+    case caf::sec::invalid_downstream:
+    case caf::sec::no_downstream_stages_defined:
+    case caf::sec::stream_init_failed:
+    case caf::sec::invalid_stream_state:
+    case caf::sec::unhandled_stream_error:
+    case caf::sec::bad_function_call:
+    case caf::sec::feature_disabled:
+    case caf::sec::cannot_open_file:
+    case caf::sec::socket_invalid:
+    case caf::sec::socket_disconnected:
+    case caf::sec::socket_operation_failed:
+    case caf::sec::unavailable_or_would_block:
+    case caf::sec::malformed_basp_message:
+    case caf::sec::serializing_basp_payload_failed:
+    case caf::sec::redundant_connection:
+    case caf::sec::remote_lookup_failed:
+    case caf::sec::no_tracing_context:
+    case caf::sec::all_requests_failed:
+    case caf::sec::field_invariant_check_failed:
+    case caf::sec::field_value_synchronization_failed:
+    case caf::sec::invalid_field_type:
+    case caf::sec::unsafe_type:
+    case caf::sec::save_callback_failed:
+    case caf::sec::load_callback_failed:
+    case caf::sec::conversion_failed:
+    case caf::sec::connection_closed:
+    case caf::sec::type_clash:
+    case caf::sec::unsupported_operation:
+    case caf::sec::no_such_key:
+    case caf::sec::broken_promise:
+    case caf::sec::connection_timeout:
+    case caf::sec::action_reschedule_failed:
+      return true;
+    case caf::sec::incompatible_versions:
+    case caf::sec::incompatible_application_ids:
+    case caf::sec::request_timeout:
+      return false;
+  }
+  return false;
+}
+
+bool is_recoverable_error(const caf::error& err) {
+  if (err.category() != caf::type_id_v<caf::sec>)
+    return true;
+  const auto err_code = std::underlying_type_t<caf::sec>{err.code()};
+  auto err_enum = caf::sec{caf::sec::none};
+  if (!caf::from_integer(err_code, err_enum)) {
+    VAST_WARN("unable to retrieve error code for a remote node connection "
+              "error:{}",
+              err);
+    return true;
+  }
+  return is_recoverable_error_enum(err_enum);
+}
+
+std::optional<caf::timespan> calculate_remaining_time(
+  const std::optional<std::chrono::steady_clock::time_point>& deadline) {
+  if (!deadline)
+    return caf::infinite;
+  const auto now = std::chrono::steady_clock::now();
+  if (now >= *deadline)
+    return std::nullopt;
+  return *deadline - now;
+}
+
+bool should_retry(const caf::error& err,
+                  const std::optional<caf::timespan>& remaining_time,
+                  caf::timespan delay) {
+  return remaining_time && *remaining_time > delay && is_recoverable_error(err);
+}
+
+std::string format_time(caf::timespan timespan) {
+  if (caf::is_infinite(timespan))
+    return "infinite";
+  std::vector<std::string> measurements;
```

Comment:
We have a similar formatter for `duration` (which is an alias for `caf::timespan`) when it is wrapped inside a `data`, so this function can also be written like this:

```cpp
std::string format_time(caf::timespan timespan) {
   if (caf::is_infinite(timespan))
     return "infinite";
  return fmt::to_string(data{timespan});
}
```

Definitely :+1: for making sure that the timespans are user-readable in log messages, that's really nice to see.

Reply:
Seems that real_printer provide wrong output. For example 1.9998 will be printer as 1.1
It multiplies the 0.998 * 100 (MaxDigits * 10) resulting in 99.8. Then it uses std::round which results in 100. The 100 is then translated to 0.1 part. Should i try to fix it as a part of this PR? I have found a nice C++ 17 function that can simplify the code here (std::to_chars).

Reply:
I don't think that it should go into this, but please create an issue so we can track this. I don't think `std::to_chars` is available on all platforms we support.